realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.kt

```diff
@@ -2066,107 +2066,126 @@ class RealmProxyClassGenerator(private val processingEnvironment: ProcessingEnvi
-            }
-            emitEmptyLine()
-            emitStatement("final %1\$s objProxy = (%1\$s) obj", interfaceName)
-            for (field in metadata.fields) {
-                val fieldName = field.simpleName.toString()
-                val qualifiedFieldType = QualifiedClassName(field.asType().toString())
-                if (metadata.isPrimaryKey(field)) {
-                    continue  // Primary key has already been set when adding new row or finding the existing row.
+                    beginControlFlow("if (obj == null)")
+                        buildExcludeFieldsList(writer, metadata.fields)
+                        val primaryKeyFieldType = QualifiedClassName(metadata.primaryKey!!.asType().toString())
+                        val primaryKeyFieldName = metadata.primaryKey!!.simpleName.toString()
+                        RealmJsonTypeHelper.emitCreateObjectWithPrimaryKeyValue(qualifiedJavaClassName, generatedClassName, primaryKeyFieldType, primaryKeyFieldName, writer)
+                    endControlFlow()
                 }
-                when {
-                    Utils.isRealmModel(field) -> RealmJsonTypeHelper.emitFillRealmObjectWithJsonValue(
-                            "objProxy",
-                            metadata.getInternalSetter(fieldName),
-                            fieldName,
-                            qualifiedFieldType,
-                            Utils.getProxyClassSimpleName(field),
-                            writer)
-                    Utils.isRealmModelList(field) -> RealmJsonTypeHelper.emitFillRealmListWithJsonValue(
-                            "objProxy",
-                            metadata.getInternalGetter(fieldName),
-                            metadata.getInternalSetter(fieldName),
-                            fieldName,
-                            (field.asType() as DeclaredType).typeArguments[0].toString(),
-                            Utils.getProxyClassSimpleName(field),
-                            writer)
-                    Utils.isRealmValueList(field) -> emitStatement("ProxyUtils.setRealmListWithJsonObject(objProxy.%1\$s(), json, \"%2\$s\")", metadata.getInternalGetter(fieldName), fieldName)
-                    Utils.isMutableRealmInteger(field) -> RealmJsonTypeHelper.emitFillJavaTypeWithJsonValue(
-                            "objProxy",
-                            metadata.getInternalGetter(fieldName),
-                            fieldName,
-                            qualifiedFieldType,
-                            writer)
-                    else -> RealmJsonTypeHelper.emitFillJavaTypeWithJsonValue(
-                            "objProxy",
-                            metadata.getInternalSetter(fieldName),
-                            fieldName,
-                            qualifiedFieldType,
-                            writer)
+                emitEmptyLine()
+                emitStatement("final %1\$s objProxy = (%1\$s) obj", interfaceName)
```

Comment:
@cmelchior Any particular reason for this to be the interface type? 

Adding the embedded objects require a `RealmModel` parent. Currently the fillers just use a hardcoded `obj` reference, but could of course just add both the interface and implementation typed variables as arguments  to the fillers where needed.

Reply:
Yes, I guess we now need both the proxy and the actual object (`RealmModel`)

Reply:
The proxy is always a subclass of the realm model, so you can do unsafe casts if needed. We do that in a few places.

Reply:
Did the unsafe cast instead, as I think it will cause less confusion than juggling differently typed references to the same object.